internals/overlord/logstate/loki/loki_test.go

```diff
@@ -191,6 +191,65 @@ func (*suite) TestServerTimeout(c *C) {
 	c.Assert(err, ErrorMatches, ".*context deadline exceeded.*")
 }
 
+func (*suite) TestBufferFull(c *C) {
+	client := loki.NewClientWithOptions(
+		&plan.LogTarget{
+			Name:     "tgt1",
+			Location: "fake",
+		},
+		&loki.ClientOptions{
+			MaxRequestEntries: 3,
+		},
+	)
+
+	checkBuffer := func(expected []any) {
+		checkBuffer(c, client, expected)
+	}
+	addEntry := func(s string) {
+		err := client.Add(servicelog.Entry{Message: s})
+		c.Assert(err, IsNil)
+	}
+
+	checkBuffer([]any{nil, nil, nil, nil, nil, nil})
+	addEntry("1")
+	checkBuffer([]any{"1", nil, nil, nil, nil, nil})
+	addEntry("2")
+	checkBuffer([]any{"1", "2", nil, nil, nil, nil})
+	addEntry("3")
+	checkBuffer([]any{"1", "2", "3", nil, nil, nil})
+	addEntry("4")
+	checkBuffer([]any{nil, "2", "3", "4", nil, nil})
+	addEntry("5")
+	checkBuffer([]any{nil, nil, "3", "4", "5", nil})
+	addEntry("6")
+	checkBuffer([]any{nil, nil, nil, "4", "5", "6"})
+	addEntry("7")
+	checkBuffer([]any{"5", "6", "7", nil, nil, nil})
+}
+
+// Check if the client's buffer is as expected
+func checkBuffer(c *C, client *loki.Client, expected []any) {
+	buffer := loki.GetBuffer(client)
+	if len(buffer) != len(expected) {
+		c.Fatalf("buffer length is %v, expected %v", len(buffer), len(expected))
+	}
+
+	for i := range expected {
+		// 'nil' means c.buffer[i] should be zero
+		if expected[i] == nil {
+			c.Assert(buffer[i], DeepEquals, loki.LokiEntryWithService{},
+				Commentf("buffer[%d] should be zero, obtained %v", i, buffer[i]))
+
+		} else if msg, ok := expected[i].(string); ok {
```

Comment:
To simplify, we can probably just use a plain `else` and do `msg := expected[i].(string)` and let it panic (programmer error in the test).

Alternatively we could use `[]string` instead of `[]any` and test for empty string instead of nil -- bit more type safe and simpler?

Reply:
I think the nil makes it clear that the entry should be zero, especially with syntax highlighting. You're right that it's probably fine to panic here though